routes/project/user.rb

```diff
@@ -176,15 +181,67 @@ class Clover
       r.post "policy/managed" do
         authorize("Project:user", @project.id)
 
-        invitation_policies = r.params["invitation_policies"] || {}
-        allowed_tags = dataset_authorize(@project.subject_tags_dataset, "SubjectTag:add").select_hash(:name, Sequel.as(true, :v))
-        invitation_policies.each do |email, policy|
-          if allowed_tags[policy]
-            @project.invitations.find { _1.email == email }&.update(policy: policy)
+        DB.transaction do
+          allowed_add_tags = dataset_authorize(@project.subject_tags_dataset, "SubjectTag:add").all.to_h { [_1.name, _1] }
+          allowed_remove_tags = dataset_authorize(@project.subject_tags_dataset, "SubjectTag:remove").all.to_h { [_1.name, _1] }
+          removals = Hash.new(0)
+          additions = Hash.new(0)
+          changes = []
+          issues = []
+          user_policies = r.params["user_policies"] || {}
+          user_policies.each do |ubid, policy|
+            policy = nil if policy == ""
+            user = Account.from_ubid(ubid)
+            existing_tags = user.subject_tags_dataset.where(project_id: @project.id).select_map(:name)
+            next unless existing_tags.length < 2
+            next if existing_tags.include?(policy)
+            existing_tags.each do |tag|
+              unless (removed_tag = allowed_remove_tags[tag])
+                issues << "You don't have permission to remove members from '#{tag}' tag"
+                next
+              end
+              removed_tag.remove_members(user.id)
```

Comment:
This approach uses a query per user.  A more optimized approach would be to collect all users to add and remove, and issue one add and/or delete query per tag to be modified, for all users that would be affected.  The tag code will insert/delete multiple rows in a single query if you call add_members/remove_members with an array of multiple arguments.

Additionally, I don't think we should remove the user from the tag unless we can add them to the new tag.

I recommend an approach where you only record the adds or removes that should be made during the loop, and then apply all changes after the loop.  You can make the `removals` and `additions` values arrays, append them with the member ids during the loop, and after the loop, loop over `removals` and call `remove_members` once for non-empty values, and them loop over `additions` and call `add_members` once for non-empty values.  You can get the counts by calling `size` on the arrays.

Reply:
I've refactored it an now writing tests to increase coverage

Reply:
The coverage is now 100% 🚀